src/package-operations.ts

```diff
@@ -9,97 +9,223 @@ import {
+   * linear ordering of its vertices such that for every directed edge uv from
+   * vertex u to vertex v, u comes before v in the ordering."
+   * Ref: https://en.wikipedia.org/wiki/Topological_sorting
+   *
+   * In our usage, there is a directed edge from PackageA to PackageB if
+   * PackageA lists PackageB in any of the dependency fields in its
+   * `package.json`.
+   *
+   * `overallOrder` is a convention of the `dependency-graph` library, which by
+   * all appearances is the opposite of topological order.
+   * Ref: https://github.com/jriecken/dependency-graph/tree/072cdcc3eef1ee4531114e7a675e3cbe828fb33e#examples
+   */
+
+  // Traverse the monorepo's internal dependency graph in topological order.
+  return dependencyGraph
+    .overallOrder() // leaves -> root
+    .reverse() // root -> leaves (topological order)
+    .reduce((fullPackageMetadata, packageName) => {
+      const shouldUpdate = shouldUpdateMonorepoPackage(
+        packageName,
+        packagesToUpdate,
+        dependencyGraph,
+        versionSyncStrategy,
+      );
+
+      // Update the set of packages that will be updated in preparation for the
+      // next iteration.
+      if (shouldUpdate) {
+        packagesToUpdate.add(packageName);
+      }
+
+      fullPackageMetadata.set(packageName, {
+        // Every package is in the dependency graph, so we know that the key exists.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        ...intermediatePackageMetadata.get(packageName)!,
+        shouldUpdate,
+      });
+      return fullPackageMetadata;
+    }, new Map<string, PackageMetadata>());
+}
+
+/**
+ * **ATTN:** This function must be called on each monorepo package in
+ * topological order of dependency in order to return valid results.
+ *
+ * Gets whether the specified package should be updated, given the set of
+ * dependents of this package that will be updated, the monorepo dependency
+ * graph, and the active version synchronization strategy.
+ *
+ * @param packageName - The name of the package to check.
+ * @param packagesToUpdate - The (mutable) set of packages that will be updated.
+ * @param dependencyGraph - The monorepo's internal dependency graph.
+ * @param versionSyncStrategy - The version synchronization strategy.
+ * @returns Whether the package should be updated to the new version.
+ */
+export function shouldUpdateMonorepoPackage(
+  packageName: string,
+  packagesToUpdate: Set<string>,
+  dependencyGraph: DepGraph<string>,
+  versionSyncStrategy: VersionSynchronizationStrategies,
+): boolean {
+  const shouldUpdate =
+    versionSyncStrategy === VersionSynchronizationStrategies.fixed ||
+    packagesToUpdate.has(packageName) ||
+    dependencyGraph
+      .dependenciesOf(packageName)
+      .reduce<boolean>((hasChangedDependency, dependencyName) => {
+        return hasChangedDependency || packagesToUpdate.has(dependencyName);
+      }, false);
```

Comment:
It appears that both `transitive` and `independent` exhibit the same behavior here, namely, that a package is updated then those updates cascade downstream. If that is not what we want, then should we say this?

``` suggestion
    (
      versionSyncStrategy !== VersionSynchronizationStrategies.independent &&
      dependencyGraph
        .dependenciesOf(packageName)
        .some((dependencyName) => packagesToUpdate.has(dependencyName))
    );
```